File: ios/SpentyAI/SpentyAI/Features/Reconciliation/ReconciliationViewModel.swift
import Foundation
import Observation

@Observable
final class ReconciliationViewModel {

    // MARK: - Data

    var statements: [Statement] = []
    var accounts: [Account] = []
    var accountSubTypes: [AccountSubType] = []
    var categories: [Category] = []
    var activeStatement: Statement?
    var parsedEntries: [ParsedEntry] = []
    var reconciliationResult: ReconciliationResult?

    // MARK: - Upload State

    var selectedSubType: String = ""
    var selectedAccountId: String = ""
    var periodFrom: Date = Date()
    var periodTo: Date = Date()
    var isUploading: Bool = false
    var uploadProgress: Double = 0

    // MARK: - UI State

    var isLoading: Bool = false
    var isReconciling: Bool = false
    var isProcessing: Bool = false
    var showUploadSheet: Bool = false
    var showUnlockSheet: Bool = false
    var unlockPassword: String = ""
    var errorMessage: String?
    var successMessage: String?

    // MARK: - Polling

    private var parsePollingTimer: Timer?

    func startParsePollingIfNeeded(statementId: String) {
        guard activeStatement?.status?.lowercased() == "parsing" else { return }
        stopParsePolling()
        parsePollingTimer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { [weak self] _ in
            guard let self else { return }
            Task { @MainActor in
                await self.pollStatement(id: statementId)
            }
        }
    }

    func stopParsePolling() {
        parsePollingTimer?.invalidate()
        parsePollingTimer = nil
    }

    @MainActor
    private func pollStatement(id: String) async {
        do {
            let updated = try await repository.fetchStatement(id: id)
            activeStatement = updated
            reconciliationResult = updated.reconciliation
            if updated.status?.lowercased() != "parsing" {
                stopParsePolling()
                // Refresh entries now that parsing is done
                parsedEntries = try await repository.fetchEntries(statementId: id)
                updateStatementInList(updated)
            }
        } catch {
            // Polling errors are non-fatal; stop polling on repeated failures
        }
    }

    // MARK: - Private

    private let repository = ReconciliationRepository.shared

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    // MARK: - Filtered Accounts

    var filteredAccounts: [Account] {
        guard !selectedSubType.isEmpty else { return accounts }
        return accounts.filter { $0.subType?.lowercased() == selectedSubType.lowercased() }
    }

    // MARK: - Load

    @MainActor
    func loadInitial() async {
        isLoading = true
        errorMessage = nil

        // Load each resource independently so one failure doesn't block the rest
        async let stmts = repository.fetchStatements()
        async let accts = repository.fetchAccounts()
        async let subTypes = repository.fetchAccountSubTypes()
        async let cats = repository.fetchCategories()

        do { statements = try await stmts } catch {
            errorMessage = error.localizedDescription
        }
        do { accounts = try await accts } catch { /* non-fatal */ }
        do { accountSubTypes = try await subTypes } catch { /* non-fatal */ }
        do { categories = try await cats } catch { /* non-fatal */ }

        isLoading = false
    }

    @MainActor
    func refreshStatements() async {
        errorMessage = nil
        do {
            statements = try await repository.fetchStatements()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Upload

    @MainActor
    func uploadStatement(fileData: Data, filename: String) async {
        guard !selectedAccountId.isEmpty else {
            errorMessage = "Please select an account."
            return
        }
        isUploading = true
        errorMessage = nil
        let mimeType = filename.lowercased().hasSuffix(".pdf") ? "application/pdf" : "text/csv"
        do {
            let response = try await repository.uploadStatement(
                fileData: fileData,
                filename: filename,
                mimeType: mimeType,
                accountId: selectedAccountId,
                periodFrom: Self.dateFormatter.string(from: periodFrom),
                periodTo: Self.dateFormatter.string(from: periodTo)
            )
            // Backend returns a lightweight response; try to refetch the full statement
            // but don't fail the whole upload if the refetch/decode fails
            if let stmtId = response.statementId {
                if let fullStatement = try? await repository.fetchStatement(id: stmtId) {
                    statements.insert(fullStatement, at: 0)
                } else {
                    // Refetch failed (likely decode issue) — refresh the whole list instead
                    await refreshStatements()
                }
            } else {
                // No statement ID returned — refresh the whole list
                await refreshStatements()
            }
            showUploadSheet = false
            successMessage = "Statement uploaded successfully."
            resetUploadForm()
        } catch {
            errorMessage = error.localizedDescription
        }
        isUploading = false
    }

    private func resetUploadForm() {
        selectedSubType = ""
        selectedAccountId = ""
        periodFrom = Date()
        periodTo = Date()
    }

    // MARK: - Statement Detail

    @MainActor
    func loadStatement(id: String) async {
        isLoading = true
        errorMessage = nil

        // Load statement and entries separately so one failure doesn't block the other
        do {
            activeStatement = try await repository.fetchStatement(id: id)
            reconciliationResult = activeStatement?.reconciliation
        } catch {
            errorMessage = error.localizedDescription
        }

        do {
            parsedEntries = try await repository.fetchEntries(statementId: id)
        } catch {
            // Non-fatal: statement header may still display
            if errorMessage == nil {
                errorMessage = "Failed to load entries: \(error.localizedDescription)"
            }
        }

        isLoading = false

        // Start polling if the statement is currently being parsed
        startParsePollingIfNeeded(statementId: id)
    }

    // MARK: - Reconcile

    @MainActor
    func reconcile(statementId: String) async {
        isReconciling = true
        errorMessage = nil
        do {
            _ = try await repository.reconcile(statementId: statementId)
            // Backend returns raw reconcile results; refetch the full statement
            let updated = try await repository.fetchStatement(id: statementId)
            activeStatement = updated
            reconciliationResult = updated.reconciliation
            updateStatementInList(updated)
            successMessage = "Reconciliation complete."
        } catch {
            errorMessage = error.localizedDescription
        }
        isReconciling = false
    }

    // MARK: - Add Missing to Ledger

    @MainActor
    func addMissingToLedger(statementId: String, entryIndices: [Int]? = nil) async {
        isProcessing = true
        errorMessage = nil
        do {
            // If no indices specified, send all missing entry indices
            let indices = entryIndices ?? {
                guard let missing = reconciliationResult?.missing, missing > 0 else { return [Int]() }
                return Array(0..<missing)
            }()
            let response = try await repository.addMissingToLedger(statementId: statementId, entryIndices: indices)
            // Refetch the full statement to update UI
            let updated = try await repository.fetchStatement(id: statementId)
            activeStatement = updated
            updateStatementInList(updated)
            successMessage = response.message ?? "Missing entries added to ledger."
        } catch {
            errorMessage = error.localizedDescription
        }
        isProcessing = false
    }

    // MARK: - Re-audit

    @MainActor
    func reaudit(statementId: String) async {
        isProcessing = true
        errorMessage = nil
        do {
            _ = try await repository.reaudit(statementId: statementId)
            // Backend returns lightweight response; refetch full statement
            let updated = try await repository.fetchStatement(id: statementId)
            activeStatement = updated
            reconciliationResult = updated.reconciliation
            updateStatementInList(updated)
            successMessage = "Re-audit started."
        } catch {
            errorMessage = error.localizedDescription
        }
        isProcessing = false
    }

    // MARK: - Unlock

    @MainActor
    func unlock(statementId: String) async {
        guard !unlockPassword.isEmpty else {
            errorMessage = "Please enter the PDF password."
            return
        }
        isProcessing = true
        errorMessage = nil
        do {
            _ = try await repository.unlock(statementId: statementId, password: unlockPassword)
            // Backend returns lightweight response; refetch full statement
            let updated = try await repository.fetchStatement(id: statementId)
            activeStatement = updated
            parsedEntries = updated.parsedEntries ?? []
            updateStatementInList(updated)
            showUnlockSheet = false
            unlockPassword = ""
            successMessage = "Statement unlocked and parsing."
        } catch {
            errorMessage = error.localizedDescription
        }
        isProcessing = false
    }

    // MARK: - Entry Category Update

    @MainActor
    func updateEntryCategory(statementId: String, index: Int, categoryId: String?, categoryName: String?) async {
        errorMessage = nil
        do {
            let response = try await repository.updateEntry(
                statementId: statementId,
                index: index,
                categoryId: categoryId,
                categoryName: categoryName
            )
            // Update the entry in-place if we got one back
            if let updatedEntry = response.entry, let entryIndex = response.entryIndex, entryIndex < parsedEntries.count {
                parsedEntries[entryIndex] = updatedEntry
            }
            // Also refetch the full statement to stay in sync
            let updated = try await repository.fetchStatement(id: statementId)
            activeStatement = updated
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Bulk Categorize

    @MainActor
    func bulkCategorize(statementId: String, categoryId: String) async {
        isProcessing = true
        errorMessage = nil
        do {
            // Build updates array for all current entries
            let updates = parsedEntries.enumerated().map { index, _ in
                BulkCategorizeEntry(entryIndex: index, categoryId: categoryId, subcategoryId: nil)
            }
            _ = try await repository.bulkCategorize(statementId: statementId, updates: updates)
            // Refetch the full statement to get updated entries
            let updated = try await repository.fetchStatement(id: statementId)
            activeStatement = updated
            parsedEntries = updated.parsedEntries ?? parsedEntries
            successMessage = "All entries categorized."
        } catch {
            errorMessage = error.localizedDescription
        }
        isProcessing = false
    }

    // MARK: - Approve / Reject

    @MainActor
    func approveStatement(id: String) async {
        isProcessing = true
        errorMessage = nil
        do {
            _ = try await repository.approveStatement(id: id)
            // Backend returns message response; refetch the full statement
            let updated = try await repository.fetchStatement(id: id)
            activeStatement = updated
            updateStatementInList(updated)
            successMessage = "Statement approved."
        } catch {
            errorMessage = error.localizedDescription
        }
        isProcessing = false
    }

    @MainActor
    func rejectStatement(id: String) async {
        isProcessing = true
        errorMessage = nil
        do {
            _ = try await repository.rejectStatement(id: id)
            // Backend returns message response; refetch the full statement
            let updated = try await repository.fetchStatement(id: id)
            activeStatement = updated
            updateStatementInList(updated)
            successMessage = "Statement rejected."
        } catch {
            errorMessage = error.localizedDescription
        }
        isProcessing = false
    }

    // MARK: - Delete

    @MainActor
    func deleteStatement(id: String) async {
        errorMessage = nil
        do {
            _ = try await repository.deleteStatement(id: id)
            statements.removeAll { $0.id == id }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Helpers

    func accountName(for id: String?) -> String {
        guard let id else { return "Unknown" }
        return accounts.first(where: { $0.id == id })?.name ?? "Unknown"
    }

    func categoryName(for id: String?) -> String {
        guard let id else { return "Uncategorized" }
        return categories.first(where: { $0.id == id })?.name ?? "Uncategorized"
    }

    func subcategories(for categoryId: String?) -> [Category] {
        guard let categoryId else { return [] }
        return categories.first(where: { $0.id == categoryId })?.children ?? []
    }

    private func updateStatementInList(_ statement: Statement) {
        if let idx = statements.firstIndex(where: { $0.id == statement.id }) {
            statements[idx] = statement
        }
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/reconciliation/ReconciliationViewModel.kt
package com.spentyai.app.features.reconciliation

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.spentyai.app.core.models.*
import com.spentyai.app.core.network.ApiResult
import kotlinx.coroutines.*
import java.text.SimpleDateFormat
import java.util.*

class ReconciliationViewModel(private val repository: ReconciliationRepository) : ViewModel() {

    // Data
    var statements by mutableStateOf<List<Statement>>(emptyList()); private set
    var accounts by mutableStateOf<List<Account>>(emptyList()); private set
    var accountSubTypes by mutableStateOf<List<AccountSubType>>(emptyList()); private set
    var categories by mutableStateOf<List<Category>>(emptyList()); private set
    var activeStatement by mutableStateOf<Statement?>(null); private set
    var parsedEntries by mutableStateOf<List<ParsedEntry>>(emptyList()); private set
    var reconciliationResult by mutableStateOf<ReconciliationResult?>(null); private set

    // Upload State
    var selectedSubType by mutableStateOf("")
    var selectedAccountId by mutableStateOf("")
    var periodFrom by mutableStateOf(Date())
    var periodTo by mutableStateOf(Date())
    var isUploading by mutableStateOf(false); private set

    // UI State
    var isLoading by mutableStateOf(false); private set
    var isReconciling by mutableStateOf(false); private set
    var isProcessing by mutableStateOf(false); private set
    var showUploadSheet by mutableStateOf(false)
    var showUnlockSheet by mutableStateOf(false)
    var unlockPassword by mutableStateOf("")
    var errorMessage by mutableStateOf<String?>(null)
    var successMessage by mutableStateOf<String?>(null)

    // Polling
    private var pollingJob: Job? = null

    val filteredAccounts: List<Account>
        get() = if (selectedSubType.isEmpty()) accounts
        else accounts.filter { it.subType?.lowercase() == selectedSubType.lowercase() }

    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    // Load
    fun loadInitial() {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null

            val stmts = async { repository.fetchStatements() }
            val accts = async { repository.fetchAccounts() }
            val subs = async { repository.fetchAccountSubTypes() }
            val cats = async { repository.fetchCategories() }

            when (val r = stmts.await()) {
                is ApiResult.Success -> statements = r.data.statements
                is ApiResult.Failure -> errorMessage = r.error.message
            }
            (accts.await() as? ApiResult.Success)?.let { accounts = it.data.accounts }
            (subs.await() as? ApiResult.Success)?.let { accountSubTypes = it.data }
            (cats.await() as? ApiResult.Success)?.let { categories = it.data }

            isLoading = false
        }
    }

    fun refreshStatements() {
        viewModelScope.launch {
            errorMessage = null
            when (val r = repository.fetchStatements()) {
                is ApiResult.Success -> statements = r.data.statements
                is ApiResult.Failure -> errorMessage = r.error.message
            }
        }
    }

    // Statement Detail
    fun loadStatement(id: String) {
        viewModelScope.launch {
            isLoading = true
            errorMessage = null

            when (val r = repository.fetchStatement(id)) {
                is ApiResult.Success -> {
                    activeStatement = r.data
                    reconciliationResult = r.data.reconciliation
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            when (val r = repository.fetchEntries(id)) {
                is ApiResult.Success -> parsedEntries = r.data.entries
                is ApiResult.Failure -> if (errorMessage == null) errorMessage = "Failed to load entries"
            }

            isLoading = false
            startParsePollingIfNeeded(id)
        }
    }

    private fun startParsePollingIfNeeded(statementId: String) {
        if (activeStatement?.status?.lowercase() != "parsing") return
        stopParsePolling()
        pollingJob = viewModelScope.launch {
            while (isActive) {
                delay(2000)
                when (val r = repository.fetchStatement(statementId)) {
                    is ApiResult.Success -> {
                        activeStatement = r.data
                        reconciliationResult = r.data.reconciliation
                        if (r.data.status?.lowercase() != "parsing") {
                            (repository.fetchEntries(statementId) as? ApiResult.Success)?.let {
                                parsedEntries = it.data.entries
                            }
                            updateStatementInList(r.data)
                            break
                        }
                    }
                    is ApiResult.Failure -> break
                }
            }
        }
    }

    fun stopParsePolling() {
        pollingJob?.cancel()
        pollingJob = null
    }

    // Reconcile
    fun reconcile(statementId: String) {
        viewModelScope.launch {
            isReconciling = true
            errorMessage = null

            when (val r = repository.reconcile(statementId)) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(statementId) as? ApiResult.Success)?.data
                    if (updated != null) {
                        activeStatement = updated
                        reconciliationResult = updated.reconciliation ?: ReconciliationResult(
                            summary = r.data.summary,
                            matched = r.data.matched,
                            missingFromLedger = r.data.missingFromLedger,
                            missingFromStatement = r.data.missingFromStatement,
                            conflicts = r.data.conflicts
                        )
                        updateStatementInList(updated)
                    }
                    successMessage = "Reconciliation complete."
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isReconciling = false
        }
    }

    // Add Missing
    fun addMissingToLedger(statementId: String, entryIndices: List<Int>? = null) {
        viewModelScope.launch {
            isProcessing = true
            errorMessage = null

            val indices = entryIndices ?: run {
                val count = reconciliationResult?.missingFromLedgerCount ?: 0
                if (count > 0) (0 until count).toList() else emptyList()
            }
            val body = AddMissingBody(entryIndices = indices)

            when (val r = repository.addMissingToLedger(statementId, body)) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(statementId) as? ApiResult.Success)?.data
                    if (updated != null) { activeStatement = updated; updateStatementInList(updated) }
                    successMessage = r.data.message ?: "Missing entries added to ledger."
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }

            isProcessing = false
        }
    }

    // Re-audit
    fun reaudit(statementId: String) {
        viewModelScope.launch {
            isProcessing = true
            errorMessage = null

            when (repository.reaudit(statementId)) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(statementId) as? ApiResult.Success)?.data
                    if (updated != null) {
                        activeStatement = updated
                        reconciliationResult = updated.reconciliation
                        updateStatementInList(updated)
                    }
                    successMessage = "Re-audit started."
                }
                is ApiResult.Failure -> errorMessage = "Re-audit failed"
            }

            isProcessing = false
        }
    }

    // Unlock
    fun unlock(statementId: String) {
        if (unlockPassword.isEmpty()) { errorMessage = "Please enter the PDF password."; return }
        viewModelScope.launch {
            isProcessing = true
            errorMessage = null

            when (repository.unlock(statementId, UnlockBody(password = unlockPassword))) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(statementId) as? ApiResult.Success)?.data
                    if (updated != null) {
                        activeStatement = updated
                        parsedEntries = updated.parsedEntries ?: emptyList()
                        updateStatementInList(updated)
                    }
                    showUnlockSheet = false
                    unlockPassword = ""
                    successMessage = "Statement unlocked and parsing."
                }
                is ApiResult.Failure -> errorMessage = "Unlock failed"
            }

            isProcessing = false
        }
    }

    // Entry Category Update
    fun updateEntryCategory(statementId: String, index: Int, categoryId: String?, categoryName: String?) {
        viewModelScope.launch {
            errorMessage = null
            val body = EntryUpdateBody(categoryId = categoryId, categoryName = categoryName)
            when (val r = repository.updateEntry(statementId, index, body)) {
                is ApiResult.Success -> {
                    val updatedEntry = r.data.entry
                    val entryIndex = r.data.entryIndex
                    if (updatedEntry != null && entryIndex != null && entryIndex < parsedEntries.size) {
                        parsedEntries = parsedEntries.toMutableList().apply { set(entryIndex, updatedEntry) }
                    }
                    (repository.fetchStatement(statementId) as? ApiResult.Success)?.let { activeStatement = it.data }
                }
                is ApiResult.Failure -> errorMessage = r.error.message
            }
        }
    }

    // Bulk Categorize
    fun bulkCategorize(statementId: String, categoryId: String) {
        viewModelScope.launch {
            isProcessing = true
            errorMessage = null

            val updates = parsedEntries.mapIndexed { index, _ ->
                BulkCategorizeEntry(entryIndex = index, categoryId = categoryId, subcategoryId = null)
            }
            val body = BulkCategorizeBody(updates = updates)

            when (repository.bulkCategorize(statementId, body)) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(statementId) as? ApiResult.Success)?.data
                    if (updated != null) {
                        activeStatement = updated
                        parsedEntries = updated.parsedEntries ?: parsedEntries
                    }
                    successMessage = "All entries categorized."
                }
                is ApiResult.Failure -> errorMessage = "Bulk categorize failed"
            }

            isProcessing = false
        }
    }

    // Approve / Reject
    fun approveStatement(id: String) {
        viewModelScope.launch {
            isProcessing = true
            errorMessage = null
            when (repository.approveStatement(id)) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(id) as? ApiResult.Success)?.data
                    if (updated != null) { activeStatement = updated; updateStatementInList(updated) }
                    successMessage = "Statement approved."
                }
                is ApiResult.Failure -> errorMessage = "Approve failed"
            }
            isProcessing = false
        }
    }

    fun rejectStatement(id: String) {
        viewModelScope.launch {
            isProcessing = true
            errorMessage = null
            when (repository.rejectStatement(id)) {
                is ApiResult.Success -> {
                    val updated = (repository.fetchStatement(id) as? ApiResult.Success)?.data
                    if (updated != null) { activeStatement = updated; updateStatementInList(updated) }
                    successMessage = "Statement rejected."
                }
                is ApiResult.Failure -> errorMessage = "Reject failed"
            }
            isProcessing = false
        }
    }

    // Delete
    fun deleteStatement(id: String) {
        viewModelScope.launch {
            errorMessage = null
            when (repository.deleteStatement(id)) {
                is ApiResult.Success -> statements = statements.filter { it.id != id }
                is ApiResult.Failure -> errorMessage = "Delete failed"
            }
        }
    }

    // Helpers
    fun accountName(id: String?): String {
        if (id == null) return "Unknown"
        return accounts.firstOrNull { it.id == id }?.name ?: "Unknown"
    }

    fun categoryName(id: String?): String {
        if (id == null) return "Uncategorized"
        return categories.firstOrNull { it.id == id }?.name ?: "Uncategorized"
    }

    fun resetUploadForm() {
        selectedSubType = ""
        selectedAccountId = ""
        periodFrom = Date()
        periodTo = Date()
    }

    fun dismissError() { errorMessage = null }
    fun dismissSuccess() { successMessage = null }

    private fun updateStatementInList(statement: Statement) {
        statements = statements.map { if (it.id == statement.id) statement else it }
    }

    override fun onCleared() {
        super.onCleared()
        stopParsePolling()
    }
}
